cortex-ar/src/critical_section.rs

```diff
@@ -1,30 +1,147 @@
 
-unsafe impl critical_section::Impl for SingleCoreCriticalSection {
-    unsafe fn acquire() -> critical_section::RawRestoreState {
-        // the i bit means "masked"
-        let was_active = !crate::register::Cpsr::read().i();
-        crate::interrupt::disable();
-        atomic::compiler_fence(atomic::Ordering::SeqCst);
-        was_active
+    /// Indicates the critical section was entered with interrupts on
+    pub const INT_ON: u8 = 0;
+
+    /// Indicates the critical section was entered with interrupts off
+    pub const INT_OFF: u8 = 1;
+
+    #[cfg(feature = "critical-section-single-core")]
+    unsafe impl critical_section::Impl for SingleCoreCriticalSection {
+        unsafe fn acquire() -> critical_section::RawRestoreState {
+            use core::sync::atomic;
+            // the i bit means "masked"
+            let was_active = !crate::register::Cpsr::read().i();
+            crate::interrupt::disable();
+            atomic::compiler_fence(atomic::Ordering::SeqCst);
+            if was_active {
+                INT_ON
+            } else {
+                INT_OFF
+            }
+        }
+
+        unsafe fn release(was_active: critical_section::RawRestoreState) {
+            use core::sync::atomic;
+            // Only re-enable interrupts if they were enabled before the critical section.
+            if was_active == INT_ON {
+                atomic::compiler_fence(atomic::Ordering::SeqCst);
+                // Safety: This is OK because we're releasing a lock that was
+                // entered with interrupts enabled
+                unsafe {
+                    crate::interrupt::enable();
+                }
+            }
+        }
     }
+}
+
+#[cfg(feature = "critical-section-multi-core")]
+mod multi_core {
+    struct MultiCoreCriticalSection;
+
+    critical_section::set_impl!(MultiCoreCriticalSection);
+
+    /// The default value for our spin-lock
+    pub const UNLOCKED: u32 = 0xFFFF_FFFF;
+
+    /// Indicates the critical section was entered with interrupts on, and the spin-lock unlocked
+    pub const INT_ON_UNLOCKED: u8 = 0;
+
+    /// Indicates the critical section was entered with interrupts off, and the spin-lock locked (by us)
+    pub const INT_OFF_LOCKED: u8 = 1;
+
+    /// Indicates the critical section was entered with interrupts off, and the spin-lock unlocked
+    pub const INT_OFF_UNLOCKED: u8 = 2;
+
+    pub static CORE_SPIN_LOCK: core::sync::atomic::AtomicU32 =
+        core::sync::atomic::AtomicU32::new(UNLOCKED);
+    unsafe impl critical_section::Impl for MultiCoreCriticalSection {
+        unsafe fn acquire() -> critical_section::RawRestoreState {
+            use core::sync::atomic;
+
+            // the i bit means "masked"
+            let was_active = !crate::register::Cpsr::read().i();
+            crate::interrupt::disable();
+
+            let core_id = crate::asm::core_id();
+
+            let locked_already = loop {
+                match CORE_SPIN_LOCK.compare_exchange(
```

Comment:
we are on ARM, so maybe `compare_exchange_weak` works as well? I'd need to check the A/R ref manual again, but I remember that it uses these special instructions to load and store IIRC.

Reply:
I think spuriously failure here would be bad, as we might believe we already hold the lock when it in fact we took the lock. Then we would unlock incorrectly.

But I'm happy for someone to argue that it would in fact be correct.

Reply:
I think you are right. Atomics are really complicated. It probably only works when there is a re-try on all error cases.